### ISSUE TITLE
GPU backends

### ISSUE THREAD

**Issue by mratsim** [mratsim/constantine#92]
Zero Knowledge Proofs work by handling constraints circuits with millions of gates corresponding to field operations.

Those can be executed in parallel and the full constant-time design with no branch of Constantine actually helps to avoid divergence at the GPU warp level.

Resources:
- https://github.com/o1-labs/snarky/pull/356
- https://github.com/MariusVanDerWijden/gpusnarks

**Comment by mratsim** [comment]
Reopening this to track potential GPU backends.

## Overview

For now we want to limit ourselves to backends supported by LLVM.

Another approach would be having a source code generator and use the corresponding runtime compiler, for example for OpenCL or Apple Metal.

## AMD GPUs

We can use the LLVM AMDGPU backend which is considered stable and included in all recent LLVM builds by default. https://www.llvm.org/docs/AMDGPUUsage.html

Relevant inline assembly:
- `S_ADDC_U32`: add-with-carry
- `S_SUBB_U32`: sub-with-borrow
- `S_MUL_HI_U32`: extended precision multiplication high limb
- `S_CSELECT_B32(dst, a, b)`: conditional select `a` if SCC flag set, b otherwise
- `S_CMOV_B32(dst, a)`: conditional move `a` into `dst` if SCC flag set.

See RDNA 3 ISA doc: https://www.amd.com/system/files/TechDocs/rdna3-shader-instruction-set-architecture-feb-2023_0.pdf

## Apple Metal

There is no official LLVM IR to Apple Metal backend but Apple uses a fork of LLVM.
By linking to it it might be possible to generate Metal shaders using a target triple of the form:
- `air64-apple-macos13.0`
- `air64-apple-ios16.0-macabi`

(see https://developer.apple.com/forums/thread/707695)

Metal doesn't seem to allow assembly for add-with-carry and extended precision multiplication.

## Nvidia Cuda

Backend configured and added in #210

## OpenCL

Generating OpenCL code through LLVM requires going through SPIR-V and loading the resulting kernel through ``clCreateProgramWithIL``

SPIR-V is an experimental backend starting from LLVM 15 and likely needs to be configured through ``LLVM_EXPERIMENTAL_TARGETS_TO_BUILD`` (see https://stackoverflow.com/questions/46905464/how-to-enable-a-llvm-backend, https://reviews.llvm.org/D115009 )

Alternatively there is https://github.com/KhronosGroup/SPIRV-LLVM-Translator but it would require compiling Nim in C++ mode.

Intel GPUs inline assembly:
- `ADDC`: add with carry
- `SUBB`: sub with borrow
- `MULH`: extended precision multiplication high limb
- `SEL(dst, a, b)`: conditional select `a` if predicate is set else `b`
- `MAD`(dst, a, b, c): multiply-add `dst = a*b+c` or multiply-accumulate ` dst += a*b+c`
- `MADW`(dst, a, b, c): extended precision MAD, stores the full 64-bit result

See also:
- https://github.com/intel/intel-graphics-compiler/blob/master/documentation/visa/6_instructions.md the Virtual ISA for OpenCL

ARM GPUs inline assembly:
TODO

## Other backends

Backends superceded by vendor-specific backends, in particular due to not being available in https://github.com/llvm/llvm-project/tree/main/llvm/lib/Target or not allowing inline assembly for add-with-carry and extended precision multiplication:
- OpenGL ES is the largest supported GPU backend (on all phones and desktops) but it doesn't seem easy (or documented) to generate performant computed code including inline assembly for extended-precision arithmetic.
- DirectX
- Vulkan, Vulkan requires the same steps as OpenCL (SPIR-V). But apparently Vulkan doesn't allow pointers? (comment from 2017: https://community.khronos.org/t/compiling-opencl-kernel-to-spir-v-then-use-in-vulkan/7213/2)
- WebGPU

**Comment by mratsim** [comment]
Some more investigation on the AMD backend: https://www.amd.com/content/dam/amd/en/documents/radeon-tech-docs/instruction-set-architectures/rdna3-shader-instruction-set-architecture-feb-2023_0.pdf

You have scalar and vector execution units, 8x more vector units in this example.
![spectacle_20240802_111949](https://github.com/user-attachments/assets/c5ce3028-22d1-4ddf-b526-81744bc3c311)
up to 2 scalar unit per wave (32 units). And vector add with carry does exist
![spectacle_20240802_112424](https://github.com/user-attachments/assets/9cfd0472-e14e-4d21-9848-775dee9b83b5)

What was mentioned S_ADDC_U32 was actually for scalar code
![spectacle_20240802_112658](https://github.com/user-attachments/assets/e23d037f-6467-4c25-98bc-991cf732d3de)

but we in-fact need vector code which does exist:
![spectacle_20240802_112933](https://github.com/user-attachments/assets/7be900f0-fbee-4274-b651-30afeffbf054)
![spectacle_20240802_113017](https://github.com/user-attachments/assets/11424a75-e5d3-422b-96e0-ee73e7705812)

**Comment by mratsim** [comment]
Looking at some of the AMD codegen, it might be that we can stay within LLVM IR as there were some LLVM improvements related to add with carry:
- https://github.com/ROCm/ROCm/issues/488
- https://github.com/llvm/llvm-project/issues/59232
- https://reviews.llvm.org/D138814 including @chfast comment
  - This is of particular interest: https://reviews.llvm.org/D138814#3973599
    ![image](https://github.com/user-attachments/assets/cf5b7a8e-1e0a-4f36-bb5f-e2c5e6994ab4)
    though I'm unsure what's canonical for substraction
    See https://godbolt.org/z/zfnorbvzr
    compile with `llc -march=amdgcn -mcpu=gfx900`
    ```
    define {i32, i1} @foo32(i32 %a, i32 %b) {
      %add = add i32 %a, %b
      %cmp = icmp ult i32 %add, %a
      %insert0 = insertvalue { i32, i1 } poison, i32 %add, 0
      %insert1 = insertvalue { i32, i1 } %insert0, i1 %cmp, 1
      ret {i32, i1} %insert1
    }

    declare { i32, i1 } @llvm.uadd.with.overflow.i32(i32, i32)

    define {i32, i1} @foo32_uaddo(i32 %a, i32 %b) {
      %add = add i32 %a, %b
      %uaddo = call { i32, i1 } @llvm.uadd.with.overflow.i32(i32 %a, i32 %b)
      ret { i32, i1 } %uaddo
    }
    ```

**Comment by mratsim** [comment]
Closed by #465
